File: 95/backend/internal/physics/world.go
package physics

import (
	"math"
	"sync"
)

const (
	FixedTimestep = 1.0 / 60.0
	MaxVelocity   = 500.0
	Friction      = 0.92
	Acceleration  = 1200.0
)

type Vec2 struct {
	X, Y float64
}

func (v Vec2) Add(o Vec2) Vec2  { return Vec2{v.X + o.X, v.Y + o.Y} }
func (v Vec2) Sub(o Vec2) Vec2  { return Vec2{v.X - o.X, v.Y - o.Y} }
func (v Vec2) Mul(s float64) Vec2 { return Vec2{v.X * s, v.Y * s} }
func (v Vec2) Len() float64      { return math.Sqrt(v.X*v.X + v.Y*v.Y) }
func (v Vec2) Normalize() Vec2 {
	l := v.Len()
	if l < 1e-6 {
		return Vec2{}
	}
	return v.Mul(1 / l)
}

type Input struct {
	MoveX    float64
	MoveY    float64
	Buttons  [17]bool
	Sequence uint16
}

type Entity struct {
	Pos      Vec2
	Vel      Vec2
	Radius   float64
	PlayerID int
}

type Obstacle struct {
	Min, Max Vec2
}

type World struct {
	mu        sync.Mutex
	Entities  []Entity
	Obstacles []Obstacle
	Time      float64
	Frame     uint64
	rng       *DeterministicRNG
}

type DeterministicRNG struct {
	state uint64
}

func NewDeterministicRNG(seed uint64) *DeterministicRNG {
	return &DeterministicRNG{state: seed}
}

func (r *DeterministicRNG) Next() uint32 {
	r.state = r.state*6364136223846793005 + 1442695040888963407
	return uint32(r.state >> 32)
}

func (r *DeterministicRNG) Float() float64 {
	return float64(r.Next()) / float64(^uint32(0))
}

func (r *DeterministicRNG) Range(min, max float64) float64 {
	return min + r.Float()*(max-min)
}

func NewWorld() *World {
	return &World{
		rng: NewDeterministicRNG(12345),
		Obstacles: []Obstacle{
			{Min: Vec2{-400, -300}, Max: Vec2{-350, 300}},
			{Min: Vec2{350, -300}, Max: Vec2{400, 300}},
			{Min: Vec2{-400, -300}, Max: Vec2{400, -250}},
			{Min: Vec2{-400, 250}, Max: Vec2{400, 300}},
			{Min: Vec2{-100, -100}, Max: Vec2{100, 100}},
		},
	}
}

func (w *World) AddPlayer(id int) {
	w.mu.Lock()
	defer w.mu.Unlock()
	for _, e := range w.Entities {
		if e.PlayerID == id {
			return
		}
	}
	w.Entities = append(w.Entities, Entity{
		Pos:      Vec2{float64(id*50 - 75), 0},
		Vel:      Vec2{},
		Radius:   20,
		PlayerID: id,
	})
}

func (w *World) RemovePlayer(id int) {
	w.mu.Lock()
	defer w.mu.Unlock()
	for i, e := range w.Entities {
		if e.PlayerID == id {
			w.Entities = append(w.Entities[:i], w.Entities[i+1:]...)
			return
		}
	}
}

func (w *World) Step(inputs map[int]Input) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.Frame++
	w.Time += FixedTimestep

	for playerIdx := range w.Entities {
		e := &w.Entities[playerIdx]
		input, ok := inputs[e.PlayerID]
		if !ok {
			e.Vel = e.Vel.Mul(Friction)
		} else {
			move := Vec2{input.MoveX, input.MoveY}
			if move.Len() > 1 {
				move = move.Normalize()
			}
			accel := move.Mul(Acceleration * FixedTimestep)
			e.Vel = e.Vel.Add(accel)
			e.Vel = e.Vel.Mul(Friction)
			if e.Vel.Len() > MaxVelocity {
				e.Vel = e.Vel.Normalize().Mul(MaxVelocity)
			}
		}

		delta := e.Vel.Mul(FixedTimestep)
		newPos := e.Pos.Add(delta)

		for _, obs := range w.Obstacles {
			if collides(newPos, e.Radius, obs) {
				normal := resolveCollision(e.Pos, newPos, e.Radius, obs)
				newPos = e.Pos.Add(delta.Sub(normal.Mul(delta.Dot(normal) * 1.01)))
				e.Vel = e.Vel.Sub(normal.Mul(e.Vel.Dot(normal) * 1.01))
				break
			}
		}

		e.Pos = newPos
	}
}

func collides(pos Vec2, r float64, obs Obstacle) bool {
	closest := Vec2{
		clamp(pos.X, obs.Min.X, obs.Max.X),
		clamp(pos.Y, obs.Min.Y, obs.Max.Y),
	}
	dist := pos.Sub(closest).Len()
	return dist < r
}

func resolveCollision(old, new Vec2, r float64, obs Obstacle) Vec2 {
	cx := clamp(new.X, obs.Min.X, obs.Max.X)
	cy := clamp(new.Y, obs.Min.Y, obs.Max.Y)
	dx := new.X - cx
	dy := new.Y - cy
	if dx == 0 && dy == 0 {
		if old.X < obs.Min.X {
			return Vec2{-1, 0}
		} else if old.X > obs.Max.X {
			return Vec2{1, 0}
		} else if old.Y < obs.Min.Y {
			return Vec2{0, -1}
		}
		return Vec2{0, 1}
	}
	dist := math.Sqrt(dx*dx + dy*dy)
	return Vec2{dx / dist, dy / dist}.Normalize()
}

func (w *World) GetPlayer(id int) (Entity, bool) {
	w.mu.Lock()
	defer w.mu.Unlock()
	for _, e := range w.Entities {
		if e.PlayerID == id {
			return e, true
		}
	}
	return Entity{}, false
}

func (w *World) Snapshot() []Entity {
	w.mu.Lock()
	defer w.mu.Unlock()
	out := make([]Entity, len(w.Entities))
	copy(out, w.Entities)
	return out
}

func (w *World) FrameNumber() uint64 {
	w.mu.Lock()
	defer w.mu.Unlock()
	return w.Frame
}

func clamp(v, lo, hi float64) float64 {
	if v < lo {
		return lo
	}
	if v > hi {
		return hi
	}
	return v
}

File: 95/frontend/src/physics/world.ts
export const FixedTimestep = 1.0 / 60.0;
export const MaxVelocity = 500.0;
export const Friction = 0.92;
export const Acceleration = 1200.0;

export interface Vec2 { x: number; y: number; }

export const vec2 = (x = 0, y = 0): Vec2 => ({ x, y });
export const vAdd = (a: Vec2, b: Vec2): Vec2 => ({ x: a.x + b.x, y: a.y + b.y });
export const vSub = (a: Vec2, b: Vec2): Vec2 => ({ x: a.x - b.x, y: a.y - b.y });
export const vMul = (a: Vec2, s: number): Vec2 => ({ x: a.x * s, y: a.y * s });
export const vLen = (a: Vec2): number => Math.sqrt(a.x * a.x + a.y * a.y);
export const vNormalize = (a: Vec2): Vec2 => {
  const l = vLen(a);
  if (l < 1e-6) return vec2();
  return vMul(a, 1 / l);
};
export const vDot = (a: Vec2, b: Vec2): number => a.x * b.x + a.y * b.y;

export interface Input {
  moveX: number;
  moveY: number;
  buttons: boolean[];
  sequence: number;
}

export interface Entity {
  pos: Vec2;
  vel: Vec2;
  radius: number;
  playerId: number;
}

export interface Obstacle {
  min: Vec2;
  max: Vec2;
}

export const clamp = (v: number, lo: number, hi: number): number =>
  v < lo ? lo : v > hi ? hi : v;

export class DeterministicRNG {
  private state: bigint;
  constructor(seed: number = 12345) {
    this.state = BigInt(seed);
  }
  next(): number {
    this.state = this.state * BigInt('63641366223846793005') + BigInt('1442695040888963407');
    return Number((this.state >> 32n) & 0xffffffffn) >>> 0;
  }
  float(): number {
    return this.next() / 0xffffffff;
  }
  range(min: number, max: number): number {
    return min + this.float() * (max - min);
  }
}

const collides = (pos: Vec2, r: number, obs: Obstacle): boolean => {
  const closest = vec2(
    clamp(pos.x, obs.min.x, obs.max.x),
    clamp(pos.y, obs.min.y, obs.max.y)
  );
  return vLen(vSub(pos, closest)) < r;
};

const resolveCollision = (oldPos: Vec2, newPos: Vec2, r: number, obs: Obstacle): Vec2 => {
  const cx = clamp(newPos.x, obs.min.x, obs.max.x);
  const cy = clamp(newPos.y, obs.min.y, obs.max.y);
  let dx = newPos.x - cx;
  let dy = newPos.y - cy;
  if (dx === 0 && dy === 0) {
    if (oldPos.x < obs.min.x) return vec2(-1, 0);
    if (oldPos.x > obs.max.x) return vec2(1, 0);
    if (oldPos.y < obs.min.y) return vec2(0, -1);
    return vec2(0, 1);
  }
  const dist = Math.sqrt(dx * dx + dy * dy);
  return vNormalize(vec2(dx / dist, dy / dist));
};

export const defaultObstacles: Obstacle[] = [
  { min: vec2(-400, -300), max: vec2(-350, 300) },
  { min: vec2(350, -300), max: vec2(400, 300) },
  { min: vec2(-400, -300), max: vec2(400, -250) },
  { min: vec2(-400, 250), max: vec2(400, 300) },
  { min: vec2(-100, -100), max: vec2(100, 100) },
];

export class World {
  entities: Entity[] = [];
  obstacles: Obstacle[] = defaultObstacles;
  time = 0;
  frame = 0;
  rng: DeterministicRNG;

  constructor(seed = 12345) {
    this.rng = new DeterministicRNG(seed);
  }

  addPlayer(id: number) {
    for (const e of this.entities) {
      if (e.playerId === id) return;
    }
    this.entities.push({
      pos: vec2(id * 50 - 75, 0),
      vel: vec2(),
      radius: 20,
      playerId: id,
    });
  }

  removePlayer(id: number) {
    const idx = this.entities.findIndex(e => e.playerId === id);
    if (idx >= 0) this.entities.splice(idx, 1);
  }

  step(inputs: Map<number, Input>) {
    this.frame++;
    this.time += FixedTimestep;

    for (const e of this.entities) {
      const input = inputs.get(e.playerId);
      if (!input) {
        e.vel = vMul(e.vel, Friction);
      } else {
        let move = vec2(input.moveX, input.moveY);
        if (vLen(move) > 1) move = vNormalize(move);
        const accel = vMul(move, Acceleration * FixedTimestep);
        e.vel = vAdd(e.vel, accel);
        e.vel = vMul(e.vel, Friction);
        if (vLen(e.vel) > MaxVelocity) {
          e.vel = vMul(vNormalize(e.vel), MaxVelocity);
        }
      }

      const delta = vMul(e.vel, FixedTimestep);
      let newPos = vAdd(e.pos, delta);

      for (const obs of this.obstacles) {
        if (collides(newPos, e.radius, obs)) {
          const normal = resolveCollision(e.pos, newPos, e.radius, obs);
          newPos = vAdd(e.pos, vSub(delta, vMul(normal, vDot(delta, normal) * 1.01)));
          e.vel = vSub(e.vel, vMul(normal, vDot(e.vel, normal) * 1.01));
          break;
        }
      }

      e.pos = newPos;
    }
  }

  getPlayer(id: number): Entity | null {
    return this.entities.find(e => e.playerId === id) || null;
  }

  snapshot(): Entity[] {
    return this.entities.map(e => ({ ...e, pos: { ...e.pos }, vel: { ...e.vel } }));
  }

  clone(): World {
    const w = new World();
    w.entities = this.snapshot();
    w.obstacles = [...this.obstacles];
    w.time = this.time;
    w.frame = this.frame;
    return w;
  }
}
